tapir/coop/tests/membership_resignation/test_service.py

```diff
@@ -17,5 +34,124 @@ def setUp(self) -> None:
+            MembershipResignation.ResignationType.BUY_BACK,
+            MembershipResignation.ResignationType.GIFT_TO_COOP,
+            MembershipResignation.ResignationType.TRANSFER,
+        ]
+
+        self.login_as_member_office_user()
+        for resignation in resignations:
+            share_owner = ShareOwnerFactory.create(nb_shares=2)
+            resignation = MembershipResignationFactory.create(
+                share_owner=share_owner,
+                resignation_type=resignation,
+            )
+            MembershipResignationService.update_shifts_and_shares_and_pay_out_day(
+                resignation=resignation,
+            )
+            shares_after_update = resignation.share_owner.share_ownerships
+            match resignation:
+                case MembershipResignation.ResignationType.BUY_BACK:
+                    for share in shares_after_update.all():
+                        self.assertEqual(
+                            resignation.cancellation_date.replace(
+                                day=31, month=12, year=resignation.cancellation_date.year + 3
+                            ),
+                            share.end_date,
+                        )
+                case MembershipResignation.ResignationType.GIFT_TO_COOP:
+                    for share in shares_after_update.all():
+                        self.assertEqual(share.end_date, resignation.cancellation_date)
+                case MembershipResignation.ResignationType.TRANSFER:
+                    self.assertEqual(ShareOwnership.objects.get(
+                        share_owner=resignation.transferring_shares_to,
+                        start_date=resignation.cancellation_date).count(),
+                        shares_after_update.count()
+                        )
+                    self.assertEqual(share.end_date, resignation.cancellation_date)
+
+    def test_update_shifts_shiftsAndShiftAttendance_cancelled(self):
+        tapir_user = TapirUserFactory.create()
+        share_owner = ShareOwnerFactory.create()
+        resignation = MembershipResignationFactory.create(share_owner=share_owner)
+        shift_template = ShiftTemplateFactory.create()
+        shift = ShiftFactory.create()
+        ShiftAttendance.objects.create(
+            user=tapir_user, slot=ShiftSlot.objects.filter(shift=shift).first()
+        )
+        ShiftAttendanceTemplate.objects.create(
+            user=tapir_user, slot_template=shift_template.slot_templates.first()
+        )
+        MembershipResignationService.update_shifts(
+            tapir_user=tapir_user, resignation=resignation
+        )
+        self.assertEqual(ShiftAttendanceTemplate.objects.count(), 0)
+        self.assertEqual(
+            ShiftAttendance.objects.get(user=tapir_user).state,
+            ShiftAttendance.State.CANCELLED,
+        )
+
+    def test_delete_end_dates_endDatesOfShares_removed(self):
+        share_owner = ShareOwnerFactory.create(nb_shares=1)
```

Comment:
The `ShareOwnership`s created here already have `None` as `end_date`, so the test will pass even if we don't call `delete_end_dates`. You'll need to set an end date manually.